bson/src/main/org/bson/codecs/ContainerCodecHelper.java

```diff
@@ -62,6 +66,23 @@ static Object readValue(final BsonReader reader, final DecoderContext decoderCon
         }
     }
 
+    static Codec<?> getCodec(final CodecRegistry codecRegistry, final Type type) {
+        if (type instanceof Class) {
+            return codecRegistry.get((Class<?>) type);
+        } else if (type instanceof ParameterizedType) {
+            ParameterizedType parameterizedType = (ParameterizedType) type;
+            Codec<?> rawCodec = codecRegistry.get((Class<?>) parameterizedType.getRawType());
```

Comment:
`ParameterizedType#getRawType` returns `Type`, not `Class`,  but can the raw type of a parameterized type be anything else except a class? If so, this is an unsafe cast.